### DISCUSSION TITLE
Enhancing the remark / VS Code experience

### DISCUSSION THREAD

**Discussion by remcohaszing** [remarkjs/remark#910] [General]
Upcoming month I have some spare time I would like to spend on enhancing the experience of using remark with VS Code.

I still need to look into the exact issues I’ll run into, but roughly this is my idea:

- [ ] `vscode-remark` should use [`unified-language-server`](https://github.com/unifiedjs/unified-language-server).
  - This will provide consistent behavior between language server based editor integrations.
  - This way plugins aren’t limited by any VS Code plugin runtime quirks.
- [x] `unified-language-server` should use [`unified-engine`](https://github.com/unifiedjs/unified-engine).  
  This ensures consistent behavior between the language server and CLI such as:
  - Loading of configuration files
  - Ignoring files
  - Processing files
- [x] `unified-language-server` should be updated to adhere to the same code quality standards as the rest of the unified ecosystem.
  - Up-to-date dependencies
  - Types based on JSDoc
  - ESM
- [ ] `vscode-remark` should be able to lint files.
- [ ] `vscode-remark` should be able to autofix / format files.

Any feedback is welcome

**Comment by wooorm** [comment]
Awesome! :) Might be something to do/communicate with [silvenon/vscode-mdx](https://github.com/silvenon/vscode-mdx) / [mdx-js/vscode-mdx](https://github.com/mdx-js/vscode-mdx) as well.

Related to linting (and https://github.com/remarkjs/remark-lint/pull/276), I also have some thoughts on improving the existing rules, and perhaps adding more lint rules for directives/math/mdx/etc? It’s a different discussion but perhaps adjacent.

**Comment by remcohaszing** [comment]
Here’s a sneak preview of what this looks like in VSCode. :smile:

It lints and formats the readme based on the latest version of `remark-preset-wooorm`. I don’t think VSCode extensions support ESM, but it doesn’t matter, as the language server works language agnostic. Also the entire VSCode extension is ~40 lines of code.

https://user-images.githubusercontent.com/779047/145581604-089b4855-4188-4a7d-a2ae-e9a8bdf5ff44.mp4
